packages/business-features/keyboard-shortcuts/src/invoke-shortcut.injectable.ts

```diff
@@ -0,0 +1,71 @@
+import { pipeline } from "@ogre-tools/fp";
+import { filter, isString } from "lodash/fp";
+import { getInjectable } from "@ogre-tools/injectable";
+import {
+  Binding,
+  KeyboardShortcut,
+  keyboardShortcutInjectionToken,
+} from "./keyboard-shortcut-injection-token";
+
+export type InvokeShortcut = (event: KeyboardEvent) => void;
+
+const toShortcutsWithMatchingScope = (shortcut: KeyboardShortcut) => {
+  const activeScopeElement = document.activeElement?.closest("[data-keyboard-shortcut-scope]");
+
+  if (!activeScopeElement) {
+    const shortcutIsRootLevel = !shortcut.scope;
+
+    return shortcutIsRootLevel;
+  }
+
+  const castedActiveScopeElementHtml = activeScopeElement as HTMLDivElement;
+
+  // eslint-disable-next-line xss/no-mixed-html
+  const activeScope = castedActiveScopeElementHtml.dataset.keyboardShortcutScope;
+
+  return shortcut.scope === activeScope;
+};
+
+const toBindingWithDefaults = (binding: Binding) =>
+  isString(binding)
+    ? { code: binding, shift: false, ctrl: false, altOrOption: false, meta: false }
+    : { ctrl: false, shift: false, altOrOption: false, meta: false, ...binding };
```

Comment:
Why not just have the param be `SetRequired<Partial<Binding>, "code">`?

Furthermore, should do some checks here concerning trying to make sure someone doesn't do `Ctrl+KeyK` as the `code` field?

Reply:
> Furthermore, should do some checks here concerning trying to make sure someone doesn't do Ctrl+KeyK as the code field?

We don't have to worry about it. Modifiers trigger their own event in the listener, which means that technically you can't say that the code is `Ctrl+KeyK` because that doesn't exist. However, if you wish, you are able to add keybinding to e.g. `ControlLeft`. But, I don't see that as a big deal, because that would be just developer doing weird stuff.

Reply:
> Why not just have the param be SetRequired<Partial<Binding>, "code">?

I don't understand the question here.